biz.aQute.bndlib/src/aQute/bnd/osgi/repository/CapabilityValidation.java

```diff
@@ -0,0 +1,259 @@
+package aQute.bnd.osgi.repository;
+
+import java.text.MessageFormat;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+
+import org.osgi.framework.Version;
+import org.osgi.framework.namespace.BundleNamespace;
+import org.osgi.framework.namespace.ExecutionEnvironmentNamespace;
+import org.osgi.framework.namespace.HostNamespace;
+import org.osgi.framework.namespace.IdentityNamespace;
+import org.osgi.framework.namespace.NativeNamespace;
+import org.osgi.framework.namespace.PackageNamespace;
+import org.osgi.resource.Capability;
+
+import aQute.bnd.osgi.resource.CapReqBuilder;
+
+class CapabilityValidation {
+
+	private static class Attribute<T> {
+		final String	namespace;
+		final String	name;
+		final Function<Object, T>	check;
+
+		Attribute(String namespace, String name, Function<Object, T> check) {
+			this.namespace = namespace;
+			this.name = name;
+			this.check = check;
+		}
+
+		boolean validate(Map<String, ? super T> map) throws IllegalArgumentException {
+			final boolean modified;
+			Object value = map.get(name);
+
+			final T checked;
+			try {
+				checked = check.apply(value);
+			} catch (Exception e) {
+				throw new IllegalArgumentException(
+					MessageFormat.format("Capability in namespace {0} has invalid value for {1} \"{2}\": {3}",
+						namespace, getClass().getSimpleName(), name, e.getMessage()));
+			}
+
+			if (checked != value) {
+				map.put(name, checked);
+				modified = true;
+			} else {
+				modified = false;
+			}
+			return modified;
+		}
+	}
+
+	private static class Directive extends Attribute<String> {
+		Directive(String namespace, String name, Function<Object, String> check) {
+			super(namespace, name, check);
+		}
+	}
+
+	private static final Map<String, List<Attribute<?>>>	VALIDATION_ATTRIBUTES	= new HashMap<>();
+	private static final Map<String, List<Directive>>		VALIDATION_DIRECTIVES	= new HashMap<>();
```

Comment:
I guess this is in preparation for adding validation rules for directives at some point?